### DISCUSSION TITLE
Secret Manager - policy.py LambdaMode(ServerlessExecutionMode)

### DISCUSSION THREAD

**Discussion by devan-cloud** [cloud-custodian/cloud-custodian#7689] [AWS] [answered]
Good Day All,

So my question is the following,
Does the secret manager resource currently support get_resources() when implemented as a lambda within AWS?

That does not seems to be what I have identified through the code, but my understanding of the c7n code is really limited.

What I am currently trying to test is cross-account filtering against the AWS resource secrets manager.
I am trying to do this via CloudTrail events and a Cloudwatch trigger of the Lambda.
I have been able to deploy the policy without issue, when I trigger the event I see the below output from c7n.

```
[INFO]	2022-08-25T11:38:42.842Z	adada	Found resource ids:['arn:aws:secretsmanager:eu-west-1:adad:secret:test-c7n-adad']
--
[INFO]	2022-08-25T11:38:43.167Z	d40ecad4-b83a-4a09-a81a-4638dd6a734e	Resources []
END RequestId: adada
```

From what I understand from the above output is c7n is able to find retrieve the secret's id from the event, but when it invokes get_resources, no resource is returned. So I went digging into the code to understand why and that has lead me to think the secrets_manager resource does not have get_resources implemented as of yet.

query.py

```
    def get_resources(self, ids, cache=True, augment=True):
        if not ids:
            return []
        if cache:
            resources = self._get_cached_resources(ids)
            if resources is not None:
                return resources
        try:
            resources = self.source.get_resources(ids)
            if augment:
                resources = self.augment(resources)
            return resources
        except ClientError as e:
            self.log.warning("event ids not resolved: %s error:%s" % (ids, e))
            return []
```

self.source.get_resources(ids) returns an empty list, at this point I make quite a few assumptions on where self.source is generated from, after looking at a few other resource examples, I notice they have get_resources implementations while resources/secretsmanager.py does not.

resources/iam-role.py which I have gotten working seems to have source describe mapping across to DescribeRole which contains a client that retrieves a list of resources.
```

    def get_source(self, source_type):
        if source_type in self.source_mapping:
            return self.source_mapping.get(source_type)(self)
        if source_type in sources:
            return sources[source_type](self)
        raise KeyError("Invalid Source %s" % source_type)
```

```
    @property
    def source_type(self):
        return self.data.get('source', 'describe')
```

So I am assuming secretsmanager.py requires a  describe source_mapping and describe class?
If someone out there could be kind enough to assist with confirmation on this that would be much appreciated.
I then could possibly try my hand at writing this or submit an actual enhancement issue instead of a question.

Regards,

**Comment by ajkerrigan** [comment]
Hi @devan-cloud - thanks for the report and solid troubleshooting! I think this actually bumps into an issue I ran across while working on an unrelated change (#7530). From a glance at Config's Secrets Manager support, this looks like another case where custodian's resource ID property (in this case, the secret name) doesn't match the field Config uses as `resourceId` (in this case, the secret ARN).

It may be worth extracting the `config_id`-related changes from #7530 as a dedicated PR (essentially removing my broken `json-diff` changes). And if we're lucky, adding something like `config_id = 'ARN'` to the secrets manager resource would "just work" to get your secrets manager policies working. Regardless, it would be very helpful to have more test cases!

I'm happy to have you run with this or kick it around together.

**Comment by devan-cloud** [comment, marked as answer]
Hi @ajkerrigan, 

I have identified the resolution, its quite simple and I should of kept following the code flow.
I went ahead and implemented your changes and I had the same result.
So this forced me to go a bit further and look how the resources are fetched and filtered.
And I worked out that secretsmanager.py does not instruct on using the boto3 client filtering, but uses client_filter.
This is where I realized that m.id is the filter utilized and matching that against the id within secretsmanager.py I pin pointed that the ARN cannot be used, only the name within the event.

So the resolution for this is to utilize "ids": "responseElements.name" not "ids": "responseElements.arn".

I appreciate the reply and I am sorry I can't add to the test cases. =(
